adaptive/domain.py

```diff
@@ -0,0 +1,743 @@
+        ------
+        ValueError : if x is a subdomain vertex
+        ValueError : if x is not in any subdomain
+        """
+
+    @abc.abstractmethod
+    def split_at(self, x):
+        """Split the domain at 'x'.
+
+        Removes and adds subdomains.
+
+        Returns
+        -------
+        old_subdomains : list of subdomains
+            The subdomains that were removed when splitting at 'x'.
+        new_subdomains : list of subdomains
+            The subdomains that were added when splitting at 'x'.
+
+        Raises
+        ------
+        ValueError : if x is outside of the domain or exists already
+        """
+
+    @abc.abstractmethod
+    def which_subdomains(self, x):
+        """Return the subdomains that contains 'x'.
+
+        Return
+        ------
+        subdomains : Iterable of subdomains
+            The subdomains to which 'x' belongs.
+
+        Raises
+        ------
+        ValueError : if x is outside of the domain
+        """
+
+    @abc.abstractmethod
+    def encloses(self, points):
+        """Return whether the domain encloses the points
+
+        Parameters
+        ----------
+        points : a point or sequence of points
+
+        Returns
+        -------
+        Boolean (if a single point was provided) or an array of booleans
+        if a sequence of points was provided) that is True when
+        the domain encloses the point.
+        """
+
+    @abc.abstractmethod
+    def vertices(self):
+        """Returns the vertices of the domain."""
+
+    @abc.abstractmethod
+    def neighbors(self, subdomain, n=1):
+        "Return all neighboring subdomains up to degree 'n'."
+
+    @abc.abstractmethod
+    def subdomains(self):
+        "Return all the subdomains in the domain."
+
+    @abc.abstractmethod
+    def subpoints(self, subdomain):
+        "Return all points in the interior of a subdomain."
+
+    @abc.abstractmethod
+    def clear_subdomains(self):
+        """Remove all points from the interior of subdomains.
+
+        Returns
+        -------
+        subdomains : the subdomains who's interior points were removed
+        """
+
+    @abc.abstractmethod
+    def volume(self, subdomain):
+        "Return the volume of a subdomain."
+
+    @abc.abstractmethod
+    def subvolumes(self, subdomain):
+        "Return the volumes of the sub-subdomains."
+
+
+def _choose_point_in_subinterval(a, b):
+    m = a + (b - a) / 2
+    if not a < m < b:
+        raise ValueError("{} cannot be split further".format((a, b)))
+    return m
+
+
+class Interval(Domain):
+    """A 1D domain (an interval).
+
+    Subdomains are pairs of floats (a, b).
+    """
+
+    def __init__(self, a, b):
+        if a >= b:
+            raise ValueError("'a' must be less than 'b'")
+
+        # If a sub-interval contains points in its interior, they are stored
+        # in 'sub_intervals' in a SortedList.
+        self.bounds = (a, b)
+        self.sub_intervals = dict()
+        self.points = SortedList([a, b])
```

Comment:
It would in principle be sufficient to store only the boundaries of the interval and always query for pairs of neighboring points. The current implementation is quite OK OTOH.